### DISCUSSION TITLE
Help Building .NET 8 on RHEL7

### DISCUSSION THREAD

**Discussion by n00j** [dotnet/source-build#3546] [General]
Due to some legacy reasons my company is stuck on `RHEL 7` for a little while. We have previously been using the prebuilt binaries which were working fine. Going forward it looks like .NET 8 will be dropping RHEL 7 support so I was looking to build from source.

I was hoping to get some guidance on how I would go about building .NET 8 using these instructions https://github.com/dotnet/dotnet#building

It looks to me like the prebuilt artifacts the build currently relies on also requires RHEL 8, and the build fails when I attempt to build on RHEL 7. If anyone has some pointers on how to build .NET 8 on RHEL 7, that would be very helpful. Is it possible to use .NET 7 or .NET 6 to build .NET 8 on RHEL 7?

**Comment by omajid** [comment]
I don't expect this to work easily, for a few reasons.

- Building .NET 8 will require a recent-ish pre-release of .NET 8. But as you are aware, we don't quite have a working .NET 8 on RHEL 7 to build with. You might be able to work around this by building on another distro but providing a RHEL-7-compatible glibc with a rootfs. You can then take that .NET build to RHEL 7 and then use it to build future versions of .NET on RHEL 7. The rootfs approach is supposed to work - Microsoft builds .NET this way - but it hasn't been tested in a source-build context ever, I think.
- .NET 8 requires a bunch of packages (eg, [cmake](https://github.com/dotnet/runtime/issues/74027)) with higher versions than what is available on RHEL 7. You might be able to find newer versions built by someone else. (Then again, this moves you off of the supported path on RHEL 7, if you care about that sort of thing)

> Is it possible to use .NET 7 or .NET 6 to build .NET 8 on RHEL 7?

I wish it was. It isn't :(

> Due to some legacy reasons my company is stuck on RHEL 7 for a little while.

Have you considered using a container that carries the RHEL 8 userspace but can run on RHEL 7?

**Reply by n00j** [reply]
Appreciate the response!

> Building .NET 8 will require a recent-ish pre-release of .NET 8. But as you are aware, we don't quite have a working .NET 8 on RHEL 7 to build with. You might be able to work around this by building on another distro but providing a RHEL-7-compatible glibc with a rootfs. You can then take that .NET build to RHEL 7 and then use it to build future versions of .NET on RHEL 7. The rootfs approach is supposed to work - Microsoft builds .NET this way - but it hasn't been tested in a source-build context ever, I think.

By rootfs approach, I assume you mean creating a chroot jail with the necessary dependencies? Will give this a shot, only issue I would see is I would need access to both version of glib for rhel7 (for the build of .net 8 for RHEL7), and rhel8 (for running the bootstrap sdk). 

> .NET 8 requires a bunch of packages (eg, https://github.com/dotnet/runtime/issues/74027) with higher versions than what is available on RHEL 7. You might be able to find newer versions built by someone else. (Then again, this moves you off of the supported path on RHEL 7, if you care about that sort of thing)

Funnily enough, we do this already internally so most of the dependencies required already work internally, it is only the bootstrapping sdk that doesn't run  

> Have you considered using a container that carries the RHEL 8 userspace but can run on RHEL 7?

Yes, that would be the best solution. Unfortunately we have some deployment targets that don't support container deployments, so was hoping to find a way to support those as well.

**Reply by omajid** [reply]
By rootfs, I meant the approach described here: https://github.com/dotnet/runtime/blob/main/docs/workflow/building/coreclr/cross-building.md#linux-cross-building.

You would have do this on RHEL 8 and create a RHEL 7 rootfs. I am not sure how to manually do that.

You would also need to build the rest of the SDK and associated nuget packages and then carry them over across builds. There's an example at https://github.com/ppc64le/build-scripts/blob/master/d/dotnet7/dotnet-build-ppc64le that uses a rootfs to build for ppc64le on an x86_64 machine; your scenario would be similar, just building for x86_64 (old glibc) on x86_64 (with new glibc).

**Comment by n00j** [comment]
Just an update on this in case anyone else is trying this.  I was able to create a rootfs for RHEL7 on WSL and build the .net 8 runtime successfully. Next step to build the sdk. 

The easiest way to build the rootfs I found, was to create a docker image with RHEL7 base install the necessary dependencies (build llvm from source https://github.com/llvm/llvm-project/archive/refs/tags/llvmorg-16.0.6.tar.gz), install an updated gcc via the yum package `devtoolset-12`.   
Then  flatten the image into a single layer via https://stackoverflow.com/questions/22713551/how-to-flatten-a-docker-image
Then extract out the layer into the file system using `docker save`, and extract out the tar file.

**Reply by omajid** [reply]
Nice! Would you be able to share your scripts to do this?

**Reply by n00j** [reply]
Yes, let me clean this up and I'll see if I can post that.

**Comment by n00j** [comment]
Here are the scripts I used to build the sdk for rhel7:
https://github.com/n00j/dotnet8centos7

I switched these to centos, but i'm not sure this was the best idea. Packages available in rhel7/8 yum aren't available in centos, but the general idea is there.  After running the `build.sh` script I manually combined the runtime + sdk tar.gz files and was able to run them on rhel7 to build + run a .net 8 project.

 I'm not sure on the next steps. Reading this: [building-for-new-architecture-using-a-rid-unknown-to-net](https://github.com/dotnet/source-build/blob/main/Documentation/bootstrapping-guidelines.md#building-for-new-architecture-using-a-rid-unknown-to-net), I'm supposed to now follow: [Use the cross-compiled SDK and augmented previously-source-built-archive to build a stage 1 SDK. See the Stage 1 instructions in [Building on a supported platform (Using RID known to .NET)](https://github.com/dotnet/source-build/blob/main/Documentation/bootstrapping-guidelines.md#building-on-a-supported-platform-using-rid-known-to-net).
 
 Using the --with-sdk flag, but this doesn't work as i'm not sure what `augmented previously-source-built-archive` exactly means.

**Reply by n00j** [reply]
One thing I encounter for the sdk build when I try to use it to build a project is this error:
```
MSBuild version 17.7.0-preview-23281-03+4ce2ff1f8 for .NET
  Determining projects to restore...
/sdk/sdk/8.0.100-preview.5.23303.1/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error MSB4044: The "ProcessFrameworkReferences" task was not given a value for the required parameter "RuntimeGraphPath". [/dotnettest/helloworld/helloworld.csproj]

```

I'm not sure where to configure this. If just take the sdk from the official release and combine that with my locally built runtime, it works fine.

**Reply by n00j** [reply]
In order to bypass this, I needed to pull the `Microsoft.NETCoreSdk.BundledVersions.props` and `RuntimeIdentifierGraph.json` files from the official release, and drop that into the sdk folder. I'm not sure how these file are generated.

**Reply by n00j** [reply]
I got this to work. 
1. After buliding the "bootstrap sdk" with https://github.com/n00j/dotnet8centos7, I also pull the nuget packages that are produced there in the `local-packages` folder. 
2. I extracted the sdk into a `sdk` folder, and put the nuget packages into a `packages` folder
3. Checked out a new copy of [dotnet](https://github.com/dotnet/dotnet) repo and then checked out preview 6 release. So now I have three folders `sdk`, `packages` and `dotnet`
4. Mounted these into dotnet8rhel7builder:squashed image that was produced in the bootstrap sdk 
`docker run -it --entrypoint /bin/bash -v ./packages:/packages:rw -v ./dotnet:/dotnet:rw -v ./sdk:/sdk:rw dotnet8rhel7builder:squashed `
5. Created a `.dotnet` folder in /dotnet, and copy over the /sdk folder into there
6. Ran `./prep.sh --no-sdk` 
8. Created a folder /prebuilt, and extracted out the downloaded prebuilts into there `tar -xvf /dotnet/prereqs/packages/archive/Private.SourceBuilt.Artifacts.Bootstrap.tar.gz -C /prebuilt`
9. Copy nuget packages from `/packages` into prebuilt and overwrite
10. Edit PackageVersions.props and update any versions I needed to incase the version differed from the nuget packages I copied over
11. Built and Installed openssl1.1 in /usr 
   ```bash
    cd /workdir
    wget https://ftp.openssl.org/source/openssl-1.1.1v.tar.gz
    tar -xzvf openssl-1.1.1v.tar.gz; cd openssl-1.1.1v;
    ./config --prefix=/usr --openssldir=/usr
    make install
   ```
12. In `/dotnet` ran `./build.sh --with-sdk /sdk --with-packages /prebuilt`
And this produced a source-built sdk with all the prebuilts packages for me to use to built the next version of the sdk
      -  One problem I found is the script `genmoduleindex.sh` fails for some reason when building the runtime.  I don't know what this script does, and in order to bypass this failure I needed to patch it to remove the `pipefail` option https://github.com/n00j/dotnet8centos7/blob/main/patches/runtime-patches/runtime-genmoduleindex.patch


I tested this out, and as long as I point the sdk to an openssl 1.1.1 library (via LD_LIBRARY_PATH), this runs on rhel7

**Now the problem I found is if I produce a self-contained executable, for some reason the self-contained still depends on a later version of glibc:**

```
LD_LIBRARY_PATH=/usr/local/lib64:/usr/local/lib ./hello2
./hello2: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by ./hello2)
./hello2: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by ./hello2)
```

**Does anyone know why this would be the case? And how I would go about fixing this?**

**Reply by n00j** [reply]
Ok, I figure out the issue with the self-contained binaries. If I don't specify the `RID` in the publish / build commain it defaults to `linux-x64` which targets glibc in rhel8+.

using the RID for rhel7 `rhel.7-x64` it correctly targets the correct glibc version
`LD_LIBRARY_PATH=/usr/local/lib64:/usr/local/lib ./dotnet publish -r rhel.7-x64 ../hello2/hello2.csproj  --self-contained -p:PublishSingleFile=true`

So it looks like everything is working correctly! 🎉🎉🎉 I was also able to use the preview 6 source-build sdk for rhel7 to build the preview 7 sdk successfully. 

The only question left is why I need to patch `genmoduleindex.sh`
- One problem I found is the script genmoduleindex.sh fails for some reason when building the runtime. I don't know what this script does, and in order to bypass this failure I needed to patch it to remove the `pipefail` option https://github.com/n00j/dotnet8centos7/blob/main/patches/runtime-patches/runtime-genmoduleindex.patch

**Reply by omajid** [reply]
> The only question left is why I need to patch genmoduleindex.sh

Can you add a `set -x` to the file and see what fails? It's probably this pipe https://github.com/dotnet/runtime/blob/main/eng/native/genmoduleindex.sh#L37

**Reply by n00j** [reply]
Here is the output after adding `-x`

```
 [ 81%] Generating coreclr module index file -> /dotnet/src/runtime/artifacts/source-build/self/src/artifacts/obj/coreclr/linux.x64.Release/debug/runtimeinfo/runtimemoduleindex.h
    + [[ 3 -lt 2 ]]
    + case "$(uname -s)" in
    ++ uname -s
    + cmd=/usr/local/bin/llvm-readelf
    + arg=-n
    + pattern='^[[:space:]]*Build ID: ([0-9A-Fa-f\-]+)'
    + /usr/local/bin/llvm-readelf -n /dotnet/src/runtime/artifacts/source-build/self/src/artifacts/obj/coreclr/linux.x64.Release/dlls/mscoree/coreclr/libcoreclr.so
    + read -r line
    [ 81%] Building CXX object jit/CMakeFiles/clrjit_unix_x64_x64.dir/fgprofile.cpp.o
    [ 81%] Building CXX object jit/CMakeFiles/clrjit_unix_x64_x64.dir/fgprofilesynthesis.cpp.o
    + [[ Displaying notes found in: .note.gnu.build-id =~ ^[[:space:]]*Build ID: ([0-9A-Fa-f\-]+) ]]
    + read -r line
    + [[ Owner                Data size         Description =~ ^[[:space:]]*Build ID: ([0-9A-Fa-f\-]+) ]]
    + read -r line
    + [[ GNU                  0x00000014        NT_GNU_BUILD_ID (unique build ID bitstring) =~ ^[[:space:]]*Build ID: ([0-9A-Fa-f\-]+) ]]
    + read -r line
    + [[ Build ID: 06c38c0838d36df2ca8bff31d9137086e1e6ba7f =~ ^[[:space:]]*Build ID: ([0-9A-Fa-f\-]+) ]]
    + printIdAsBinary 06c38c0838d36df2ca8bff31d9137086e1e6ba7f
    + id=06c38c0838d36df2ca8bff31d9137086e1e6ba7f
    + bytesLength=40
    + printf '0x%02x, ' 20
    + [[ -n 06c38c0838d36df2ca8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' 06
    + id=c38c0838d36df2ca8bff31d9137086e1e6ba7f
    + [[ -n c38c0838d36df2ca8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' c3
    + id=8c0838d36df2ca8bff31d9137086e1e6ba7f
    + [[ -n 8c0838d36df2ca8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' 8c
    + id=0838d36df2ca8bff31d9137086e1e6ba7f
    + [[ -n 0838d36df2ca8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' 08
    + id=38d36df2ca8bff31d9137086e1e6ba7f
    + [[ -n 38d36df2ca8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' 38
    + id=d36df2ca8bff31d9137086e1e6ba7f
    + [[ -n d36df2ca8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' d3
    + id=6df2ca8bff31d9137086e1e6ba7f
    + [[ -n 6df2ca8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' 6d
    + id=f2ca8bff31d9137086e1e6ba7f
    + [[ -n f2ca8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' f2
    + id=ca8bff31d9137086e1e6ba7f
    + [[ -n ca8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' ca
    + id=8bff31d9137086e1e6ba7f
    + [[ -n 8bff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' 8b
    + id=ff31d9137086e1e6ba7f
    + [[ -n ff31d9137086e1e6ba7f ]]
    + printf '0x%s, ' ff
    + id=31d9137086e1e6ba7f
    + [[ -n 31d9137086e1e6ba7f ]]
    + printf '0x%s, ' 31
    + id=d9137086e1e6ba7f
    + [[ -n d9137086e1e6ba7f ]]
    + printf '0x%s, ' d9
    + id=137086e1e6ba7f
    + [[ -n 137086e1e6ba7f ]]
    + printf '0x%s, ' 13
    + id=7086e1e6ba7f
    + [[ -n 7086e1e6ba7f ]]
    + printf '0x%s, ' 70
    + id=86e1e6ba7f
    + [[ -n 86e1e6ba7f ]]
    + printf '0x%s, ' 86
    + id=e1e6ba7f
    + [[ -n e1e6ba7f ]]
    + printf '0x%s, ' e1
    + id=e6ba7f
    + [[ -n e6ba7f ]]
    + printf '0x%s, ' e6
    + id=ba7f
    + [[ -n ba7f ]]
    + printf '0x%s, ' ba
    + id=7f
    + [[ -n 7f ]]
    + printf '0x%s, ' 7f
    + id=
    + [[ -n '' ]]
    + break
    make[2]: *** [debug/runtimeinfo/CMakeFiles/runtimeinfo.dir/build.make:74: debug/runtimeinfo/runtimemoduleindex.h] Error 74
    make[2]: *** Deleting file 'debug/runtimeinfo/runtimemoduleindex.h'
    make[1]: *** [CMakeFiles/Makefile2:4474: debug/runtimeinfo/CMakeFiles/runtimeinfo.dir/all] Error 2
    make[1]: *** Waiting for unfinished jobs....
    ```

**Reply by n00j** [reply]
I tried building the full release and ran into a build issue with fsharp. If anyone knowledgeable could take a look and point me to any potential issue.

```
  /dotnet/src/fsharp/artifacts/source-build/self/src/artifacts/Bootstrap/fsc/Microsoft.FSharp.Targets(331,9): error MSB6006: "dotnet" exited with code 134. [/dotnet/src/fsharp/artifacts/source-build/self/src/src/FSharp.Core/FSharp.Core.fsproj::TargetFramework=netstandard2.1]
  ##vso[task.logissue type=error;sourcepath=/dotnet/src/fsharp/artifacts/source-build/self/src/artifacts/Bootstrap/fsc/Microsoft.FSharp.Targets;linenumber=331;columnnumber=9;code=MSB6006;](NETCORE_ENGINEERING_TELEMETRY=Build) "dotnet" exited with code 134.
```
 
[fsharp.log](https://github.com/dotnet/source-build/files/13384887/fsharp.log)

**Reply by omajid** [reply]
134 generally means a SIGABRT. Can get a coredump and/or run a backtrace?

**Reply by n00j** [reply]
I got the coredump, and I opened it with gdb:

```
Missing separate debuginfo for /sdk/host/fxr/8.0.0-rc.2.23479.6/libhostfxr.so
Try: yum --enablerepo='*debug*' install /usr/lib/debug/.build-id/f0/bc45fe9254789994d8c17ee19bc7097d11c6af.debug
Missing separate debuginfo for /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libhostpolicy.so
Try: yum --enablerepo='*debug*' install /usr/lib/debug/.build-id/1a/9261c4f0c3bf5a53f26e5c67616fd7572af287.debug
Missing separate debuginfo for /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
Try: yum --enablerepo='*debug*' install /usr/lib/debug/.build-id/4c/08d0921f5508ffdd2326f5e460f0e7892e564b.debug
Missing separate debuginfo for /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclrtraceptprovider.so
Try: yum --enablerepo='*debug*' install /usr/lib/debug/.build-id/f7/3e46df5d2ed3bf1a740c77c392eb00d0bd490e.debug
Missing separate debuginfo for /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libclrjit.so
Try: yum --enablerepo='*debug*' install /usr/lib/debug/.build-id/79/ee05e7398e1d9792d2473c2d18434bf5d42ffb.debug
Missing separate debuginfo for /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libSystem.Native.so
Try: yum --enablerepo='*debug*' install /usr/lib/debug/.build-id/70/0e2dc96059f267d8900e928c6e68240b42d5f4.debug
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/sdk/dotnet /dotnet/src/fsharp/artifacts/source-build/self/src/artifacts/Bootst'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f158c586387 in raise () from /lib64/libc.so.6
[Current thread is 1 (Thread 0x7f158d770780 (LWP 13173))]
Missing separate debuginfos, use: debuginfo-install glibc-2.17-326.el7_9.x86_64 libgcc-4.8.5-44.el7.x86_64 libicu-50.2-4.el7_7.x86_64 libstdc++-4.8.5-44.el7.x86_64 lttng-ust-2.4.1-4.el7.x8--Type <RET> for more, q to quit, c to continue without paging-- c
```

**Reply by n00j** [reply]
not sure I have any debug symbols...
```
(gdb) backtrace
#0  0x00007f158c586387 in raise () from /lib64/libc.so.6
#1  0x00007f158c587a78 in abort () from /lib64/libc.so.6
#2  0x00007f158c478a47 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#3  0x00007f158c478969 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#4  0x00007f158c15d17f in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#5  0x00007f158c01e13e in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#6  0x00007f158c01e376 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#7  0x00007f158c01e4c4 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#8  0x00007f158c00f4cc in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#9  0x00007f158c00c3de in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#10 0x00007f158c02c063 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#11 0x00007f158c01a1ec in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#12 0x00007f158c03d7c2 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#13 0x00007f158c0e2757 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#14 0x00007f158c0e0f61 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#15 0x00007f158c094d1c in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#16 0x00007f158460a6be in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libclrjit.so
#17 0x00007f1584563c21 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libclrjit.so
#18 0x00007f158456279f in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libclrjit.so
#19 0x00007f158456486b in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libclrjit.so
#20 0x00007f1584569666 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libclrjit.so
#21 0x00007f158c09831b in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#22 0x00007f158c098523 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#23 0x00007f158c0990ce in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#24 0x00007f158c0d3621 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#25 0x00007f158c0d2f02 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#26 0x00007f158c0d2690 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#27 0x00007f158c0d209e in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#28 0x00007f158c045c57 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#29 0x00007f158c0d64a5 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#30 0x00007f158c0d6079 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#31 0x00007f158c2f2ea0 in ?? () from /sdk/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libcoreclr.so
#32 0x00007f15108c5612 in ?? ()
#33 0xfd25ff00003ffb15 in ?? ()
#34 0x000000012143fae8 in ?? ()
#35 0x000000000000000c in ?? ()
#36 0x00000e878d770ab0 in ?? ()
#37 0xffffffff0000000c in ?? ()
#38 0x00007ed52141fbc0 in ?? ()
#39 0x00007ed521441a88 in ?? ()
#40 0x00007ed521441a88 in ?? ()
#41 0x00007ed52141ecc8 in ?? ()
#42 0x00007ed52143fdc0 in ?? ()
#43 0x0000000121441a70 in ?? ()
#44 0x00007ed52143fae8 in ?? ()
#45 0x00007ffe76dc9d90 in ?? ()
#46 0x00007f15108c54c4 in ?? ()
#47 0x0000000000000000 in ?? ()
```

**Reply by n00j** [reply]
Is there a way to skip building fsharp? We actually don't need it atm, and if I could just skip that part for now, that would  be great.

**Reply by omajid** [reply]
> not sure I have any debug symbols...

Did you build this RC 2 SDK yourself? If so, they should have included debug symbols. If you used a prebuilt SDK from Microsoft, you can use https://learn.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-symbol and lldb to get better stack traces.

> Is there a way to skip building fsharp? 

Not really. We can remove fsharp from the build configuration, but the .NET SDK has a dependency on it. So it might pull down a version built by Microsoft from nuget.org, or might just outright fail the build.

**Comment by n00j** [comment]
> > not sure I have any debug symbols...
> 
> Did you build this RC 2 SDK yourself? If so, they should have included debug symbols. If you used a prebuilt SDK from Microsoft, you can use https://learn.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-symbol and lldb to get better stack traces.
> 
> > Is there a way to skip building fsharp?
> 
> Not really. We can remove fsharp from the build configuration, but the .NET SDK has a dependency on it. So it might pull down a version built by Microsoft from nuget.org, or might just outright fail the build.

I built the rc2 bootstrap version myself, but I might not have saved the symbols 😭 

I just put `exit 0` in the fsharp `build.sh` script and it seems to have bypassed it, and it seems to have succeeded the build. I also had to delete the prebuilt artifacts the source-build downloaded that were referencing `centos.8.*` e.g. `rm /prebuilt/*.centos.* # remove and centos 8 + packages that were downloaded`

I just tested a simple hello world on a rhel7 machine and it looks like that worked. Will do some more testing 🤞
```
> dotnet new console
> dotnet run
Hello, World!
``` 

Thanks for all the help!
